controllers/sriovoperatorconfig_controller.go

```diff
@@ -409,101 +410,103 @@ func (r *SriovOperatorConfigReconciler) syncK8sResource(cr *sriovnetworkv1.Sriov
 	return nil
 }
 
-func (r *SriovOperatorConfigReconciler) syncOffloadMachineConfig(dc *sriovnetworkv1.SriovOperatorConfig) error {
-	logger := r.Log.WithName("syncOffloadMachineConfig")
+func (r *SriovOperatorConfigReconciler) syncOvsHardwareOffloadMachineConfig(dc *sriovnetworkv1.SriovOperatorConfig) error {
+	logger := r.Log.WithName("syncOvsHardwareOffloadMachineConfig")
 	var err error
 
 	logger.Info("Start to render MachineConfig and MachineConfigPool for OVS HW offloading")
+
 	data := render.MakeRenderData()
-	data.Data["HwOffloadNodeLabel"] = HwOffloadNodeLabel
-	mcName := "00-" + HwOffloadNodeLabel
-	mcpName := HwOffloadNodeLabel
-	mc, err := render.GenerateMachineConfig("bindata/manifests/switchdev-config", mcName, HwOffloadNodeLabel, dc.Spec.EnableOvsOffload, &data)
-	if err != nil {
-		return err
-	}
-	mcpRaw, err := render.RenderTemplate("bindata/manifests/switchdev-config/machineconfigpool.yaml", &data)
+	mcpMap := make(map[string]bool)
+
+	mcpList := &mcfgv1.MachineConfigPoolList{}
+	err = r.List(context.TODO(), mcpList, &client.ListOptions{})
 	if err != nil {
-		return err
-	}
-	mcp := &mcfgv1.MachineConfigPool{}
-	if len(mcpRaw) != 1 {
-		return fmt.Errorf("Invalid MachineConfigPool CR template")
+		return fmt.Errorf("Failed to get MachineConfigPoolList: %v", err)
 	}
-	err = r.Scheme.Convert(mcpRaw[0], mcp, context.TODO())
-	if err != nil {
-		return err
+
+	for _, mcp := range mcpList.Items {
+		selector, err := metav1.LabelSelectorAsSelector(mcp.Spec.NodeSelector)
+		if err != nil {
+			return fmt.Errorf("Invalid label selector in MachineConfigPool: %s, %v", mcp.GetName(), err)
+		}
+		// MachineConfigPool is selected when its NodeSelectors are included in MachineConfigPoolSelector
+		for _, pool := range dc.Spec.OvsHardwareOffload {
+			if selector.Matches(labels.Set(pool.MachineConfigPoolSelector)) {
+				mcpMap[mcp.GetName()] = true
+				break
+			}
+		}
 	}
 
-	foundMC := &mcfgv1.MachineConfig{}
-	foundMCP := &mcfgv1.MachineConfigPool{}
+	// Offload is not supported for master MachineConfigPool
+	delete(mcpMap, "master")
 
-	err = r.Get(context.TODO(), types.NamespacedName{Name: mcName}, foundMC)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			if dc.Spec.EnableOvsOffload {
-				err = r.Create(context.TODO(), mc)
-				if err != nil {
-					return fmt.Errorf("Couldn't create MachineConfig: %v", err)
-				}
-				logger.Info("Created MachineConfig CR")
-			}
-		} else {
-			return fmt.Errorf("Failed to get MachineConfig: %v", err)
+	for mcpName, enable := range mcpMap {
+		mcName := "00-" + mcpName + "-" + OVS_HWOL_MACHINE_CONFIG_NAME_SUFFIX
```

Comment:
Do we have to create one MC for each selected pool? Have you considered if `spec.machineConfigSelector` in MCP can be used instead?

Reply:
Yes, with current API change, we probably can use single MC be matched by all MCP.
I was thinking for advanced cases that we might have HWOL configurations that differ between each MCP in the future (e.g. the tc_policy config), in which case, having one MC per MCP would be needed.